WIP – TAP_ONLY, --only, {only: true} and "only" warning

gr2m opened the pull request:
# THIS IS WORK IN PROGRESS

writes warning to stderr if {only: true} but tap’s only option is not true

closes #187 

### Progress

- [x] make existing tests pass (rebase on master https://github.com/tapjs/node-tap/pull/374#issuecomment-309129473)
- [x] create test in `test/test/...` https://github.com/tapjs/node-tap/pull/374#issuecomment-308904189
- [x] create runner test https://github.com/tapjs/node-tap/pull/374#issuecomment-308904189
- [x] add --only to the set of configs that can be loaded from a .taprc file https://github.com/tapjs/node-tap/pull/374#issuecomment-308904189
- [x] update `usage.txt`

gr2m committed:
WIP TAP_ONLY, --only, {only: true} and warning to stderr if {only: true} but tap’s only option is not true

gr2m commented:
ping @isaacs I need your input here to know if this goes in the right direction before I can continue.
If you are too busy right now let me know when / how to ping again, happy to do so

isaacs commented:
Reading through this now, sorry for the delay :)

isaacs commented on lib/test.js:
```diff
@@ -110,7 +110,11 @@ Test.prototype.spawn = function spawn (cmd, args, options, name) {
 Test.prototype.sub = function (Class, extra, caller) {
-  if (extra && (extra.todo || extra.skip)) {
+  if (extra.only && !this.only) {
+    process.stderr.write('# Warning: "' + extra.name + '" has {todo: true} set but all tests are run\n')
```
Do you mean `{only: true}`?

isaacs commented:
This strikes me as the correct approach, overall.  (Apart from what I'm pretty sure is a typo in the warning message?)

Needs a test.  You could add a js file to `test/test/...` and then `npm run regen-fixtures` to verify output.  Probably there should be a runner test as well.

It looks like we'll probably also want to add `--only` to the set of configs that can be loaded from a `.taprc` file?

isaacs commented:
Not sure why the tests seem to be hanging on node v6.  I'll try to investigate later.

gr2m committed:
WIP fix typo in warning about `{only: true}`

isaacs commented:
@gr2m If you rebase onto master, it should make the test failures go away.  `--debug` is super flaky on node 6, and removed in node 8, so I deleted the test, since it hangs about half the time.

gr2m commented:
I’ve rebased on top of master. The tests pass locally with node v6.9.5 (npm v3.10.10) and on travis with all versions now. Not sure why they fail on appveyor: https://ci.appveyor.com/project/isaacs/node-tap/build/405/job/4cfnem7oasdhwvy0

it looks like it’s not related to this PR though, as previous builds failed, too

> It looks like we'll probably also want to add --only to the set of configs that can be loaded from a .taprc file?

That seems to work already? Or am I missing something?

```bash
$ ./node_modules/.bin/tap test/only-test.js 

test/only-test.js # Warning: "only" has {only: true} set but all tests are run
test/only-test.js ..................................... 2/2
total ................................................. 2/2

  2 passing (294.111ms)

  ok

$ cat .taprc 
only: true

$ TAP_RCFILE=`pwd`/.taprc ./node_modules/.bin/tap test/only-test.js 

total ................................................. 1/1

  1 passing (7.752ms)

  ok
```

gr2m committed:
add test for `--only` flag

gr2m committed:
update usage, add `-O` alias for `--only`

ping @isaacs is the `-O` option okay? Migth be confused with the existing `-o` option for output file.

gr2m committed:
runner test for --only

gr2m commented:
> Probably there should be a runner test as well 

Is this what you had in mind? 2febf88

gr2m committed:
fixup! runner test for --only

gr2m committed:
fixup! runner test for --only

gr2m committed:
fixup! runner test for --only

gr2m commented:
This doesn’t look right yet. 

for a test file looking like this:

```js
var test = require('../..').test

test('root test without {only: true}', function (t) {
  t.pass('test will not run without --only')
  t.end()
})

test('root test with {only: true}', {only: true}, function (t) {
  t.pass('test will log an error when run without --only')
  t.end()
})

test('group', function (group) {
  group.test('subtest without {only: true}', function (t) {
    t.pass('subtest will not run without --only')
    t.end()
  })

  group.test('subtest with {only: true}', {only: true}, function (t) {
    t.pass('test will log an error when run without --only')
    t.end()
  })

  group.end()
})
```

I get this result when run without the `only` flag:

```bash
$ bin/run.js --reporter tap test/fixtures/only.js 
TAP version 13
# Subtest: test/fixtures/only.js
    # Subtest: root test without {only: true}
        ok 1 - test will not run without --only
        1..1
# Warning: "root test with {only: true}" has {only: true} set but all tests are run
    ok 1 - root test without {only: true} # time=5.29ms
    
    # Subtest: root test with {only: true}
        ok 1 - test will log an error when run without --only
        1..1
    ok 2 - root test with {only: true} # ONLY
    
    # Subtest: group
        # Subtest: subtest without {only: true}
            ok 1 - subtest will not run without --only
            1..1
        ok 1 - subtest without {only: true} # time=0.849ms
        
# Warning: "subtest with {only: true}" has {only: true} set but all tests are run
        # Subtest: subtest with {only: true}
            ok 1 - test will log an error when run without --only
            1..1
        ok 2 - subtest with {only: true} # ONLY
        
        1..2
    ok 3 - group # time=10.348ms
    
    1..3
    # time=32.296ms
ok 1 - test/fixtures/only.js # time=246.795ms

1..1
# time=260.614ms
```

And this result when run with the `only` flag

```
$ bin/run.js --only --reporter tap test/fixtures/only.js 
TAP version 13
ok 1 - test/fixtures/only.js
1..1
# time=8.261ms
```

But I would expect this output

```bash
$ bin/run.js --only --reporter tap test/fixtures/only.js 
TAP version 13
# Subtest: test/fixtures/only.js
    # Subtest: root test with {only: true}
        ok 1 - test will log an error when run without --only
        1..1
    ok 1 - root test with {only: true}
    
    # Subtest: group
        # Subtest: subtest with {only: true}
            ok 1 - test will log an error when run without --only
            1..1
        ok 1 - subtest with {only: true}
        
        1..1
    ok 2 - group # time=8.193ms
    
    1..2
    # time=30.159ms
ok 1 - test/fixtures/only.js # time=261.422ms

1..1
```

My expected output would be the same as the first, but it would leave out all tests without the `{only: true}` entirely.

Does that sound right to you @isaacs?

gr2m committed:
WIP test/fixtures/only.js with subtests

isaacs commented:
Yeah, I'd expect that the `--only` flag will set `tap.only` at the root, and then exclude any subtests that don't have `{ only: true }`, but still include sub-subtests of included tests.

Maybe you can do something like `this.only = parent ? parent.only : false` in the Test constructor?

Also, it's a little confusing to use the same flag to mean "only run this thing" and "this thing only runs child things with the 'only run this thing' flag".  Maybe consider `this.runOnly` instead of `this.only`?

isaacs commented:
@gr2m I'll have some time to review this more carefully this weekend, and I might be able to just fix the issues you're running into.

I think it's almost there, and I wanna get this and `--grep` landed before switching gears to work on v11.  Thanks for digging into it :)

isaacs commented:
@gr2m Here's what I'm talking about: https://github.com/tapjs/node-tap/commit/8a47fa238c1d70fc78cb19b2cdfe05382975d040

I haven't updated the tests, so it's probably failing, but it seems like this generates the intended result when running the fixture.

gr2m commented:
shipped in c41742440869f06ee1eac39d9fc80a3fc3af3f53. Thanks :)
